### PR TITLE
Fix dirty for loaded blocks

### PR DESCRIPTION
**Pull request by blackyblack**

### CODE REVIEW

File: include/veriblock/pop/blockchain/pop/pop_state.hpp [hunk 1]
```diff
@@ -32,27 +32,22 @@ struct PopState {
...
-  void insertContainingEndorsement(std::shared_ptr<endorsement_t> e) {
-    VBK_ASSERT_MSG(e != nullptr, "Inserted endorsement should not be nullptr");
-    _containingEndorsements.emplace(e->id, std::move(e));
-    setDirty();
-  }
-
   void insertEndorsedBy(const endorsement_t* e) {
     VBK_ASSERT_MSG(e != nullptr, "Inserted endorsement should not be nullptr");
     _endorsedBy.push_back(e);
-    setDirty();
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/pop/pop_state.hpp:44]: why?

File: include/veriblock/pop/blockchain/pop/pop_state.hpp [hunk 2]
```diff
@@ -32,27 +32,22 @@ struct PopState {
...
   bool eraseLastFromEndorsedBy(const endorsement_t* endorsement) {
     auto rm = [&endorsement](const endorsement_t* e) -> bool {
       return e == endorsement;
     };
-    auto res = erase_last_item_if<const endorsement_t*>(_endorsedBy, rm);
-    if (res) {
-      setDirty();
-    }
-    return res;
+    return erase_last_item_if<const endorsement_t*>(_endorsedBy, rm);
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/pop/pop_state.hpp:44]: why did you remove setDirty here?

File: src/pop/blockchain/btc_block_addon.cpp [hunk 3]
```diff
@@ -78,25 +78,19 @@ void BtcBlockAddon::removeRef(BtcBlockAddon::ref_height_t referencedAtHeight) {
 
 void BtcBlockAddon::insertBlockOfProofEndorsement(const VbkEndorsement* e) {
   this->_blockOfProofEndorsements.push_back(e);
-  setDirty();
```
Comment (Warchant) [inline src/pop/blockchain/btc_block_addon.cpp:81]: Same question, why?

File: src/pop/blockchain/btc_block_addon.cpp [hunk 4]
```diff
@@ -78,25 +78,19 @@ void BtcBlockAddon::removeRef(BtcBlockAddon::ref_height_t referencedAtHeight) {
...
 
 bool BtcBlockAddon::eraseLastFromBlockOfProofEndorsement(
     const VbkEndorsement* endorsement) {
   auto rm = [&endorsement](const VbkEndorsement* e) -> bool {
     return e == endorsement;
   };
-  auto res =
-      erase_last_item_if<const VbkEndorsement*>(_blockOfProofEndorsements, rm);
-  if (res) {
-    setDirty();
```
Comment (Warchant) [inline src/pop/blockchain/btc_block_addon.cpp:92]: also here

File: src/pop/blockchain/btc_block_addon.cpp [hunk 5]
```diff
@@ -78,25 +78,19 @@ void BtcBlockAddon::removeRef(BtcBlockAddon::ref_height_t referencedAtHeight) {
...
-  }
-  return res;
+  return erase_last_item_if<const VbkEndorsement*>(_blockOfProofEndorsements,
+                                                   rm);
 }
 
 void BtcBlockAddon::clearBlockOfProofEndorsement() {
   this->_blockOfProofEndorsements.clear();
-  setDirty();
```
Comment (Warchant) [inline src/pop/blockchain/btc_block_addon.cpp:99]: And here

File: src/pop/blockchain/vbk_block_addon.cpp [hunk 6]
```diff
@@ -100,25 +100,19 @@ void VbkBlockAddon::addRef(VbkBlockAddon::ref_height_t) {
 
 void VbkBlockAddon::insertBlockOfProofEndorsement(const AltEndorsement* e) {
   this->_blockOfProofEndorsements.push_back(e);
-  setDirty();
```
Comment (Warchant) [inline src/pop/blockchain/vbk_block_addon.cpp:103]: and here

File: src/pop/blockchain/vbk_block_addon.cpp [hunk 7]
```diff
@@ -100,25 +100,19 @@ void VbkBlockAddon::addRef(VbkBlockAddon::ref_height_t) {
...
 
 bool VbkBlockAddon::eraseLastFromBlockOfProofEndorsement(
     const AltEndorsement* endorsement) {
   auto rm = [&endorsement](const AltEndorsement* e) -> bool {
     return e == endorsement;
   };
-  auto res =
-      erase_last_item_if<const AltEndorsement*>(_blockOfProofEndorsements, rm);
-  if (res) {
-    setDirty();
```
Comment (Warchant) [inline src/pop/blockchain/vbk_block_addon.cpp:114]: and here

File: src/pop/blockchain/vbk_block_addon.cpp [hunk 8]
```diff
@@ -100,25 +100,19 @@ void VbkBlockAddon::addRef(VbkBlockAddon::ref_height_t) {
...
-  }
-  return res;
+  return erase_last_item_if<const AltEndorsement*>(_blockOfProofEndorsements,
+                                                  rm);
 }
 
 void VbkBlockAddon::clearBlockOfProofEndorsement() {
   this->_blockOfProofEndorsements.clear();
-  setDirty();
```
Comment (Warchant) [inline src/pop/blockchain/vbk_block_addon.cpp:121]: and here

File: include/veriblock/pop/blockchain/pop/pop_state.hpp (hunk 1 above)
Comment (blackyblack) [reply include/veriblock/pop/blockchain/pop/pop_state.hpp:44]: It is not ready for review yet. Wait a bit.

File: include/veriblock/pop/blockchain/blockchain_util.hpp [hunk 9]
```diff
@@ -111,13 +111,24 @@ bool recoverEndorsements(ProtectedBlockTree& ed_,
       auto& by = endorsed->getEndorsedBy();
       VBK_ASSERT_MSG(std::find(by.begin(), by.end(), endorsement) == by.end(),
                      "same endorsement is added to endorsedBy second time");
+      bool isDirty = endorsed->isDirty();
       endorsed->insertEndorsedBy(endorsement);
+      // keep dirty flag since recoverEndorsements is used when loading blocks from storage
+      // and should not affect dirtyness of the blocks
+      if (!isDirty) {
+        endorsed->unsetDirty();
+      }
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/blockchain_util.hpp:120]: I think you can just do `endorsed->unsetDirty()` regardless of its value. (so remove bool isDirty and if)

File: include/veriblock/pop/blockchain/blockchain_util.hpp [hunk 10]
```diff
@@ -111,13 +111,24 @@ bool recoverEndorsements(ProtectedBlockTree& ed_,
...
       const auto& bop = blockOfProof->getBlockOfProofEndorsement();
       VBK_ASSERT_MSG(
           std::find(bop.begin(), bop.end(), endorsement) == bop.end(),
           "same endorsement is added to blockOfProof second time");
+      isDirty = blockOfProof->isDirty();
       blockOfProof->insertBlockOfProofEndorsement(endorsement);
+      // keep dirty flag
+      if (!isDirty) {
+        blockOfProof->unsetDirty();
+      }
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/blockchain_util.hpp:131]: wait. it does't make sense.

if block was dirty,
 then unsetDirty won't be called.

if block was clean,
then unsetDirty will be called. 

Is this intentional?

File: include/veriblock/pop/blockchain/blockchain_util.hpp (hunk 9 above)
Comment (blackyblack) [reply include/veriblock/pop/blockchain/blockchain_util.hpp:120]: We can try it this way but I thought it is better to preserve existing state.

File: include/veriblock/pop/blockchain/blockchain_util.hpp (hunk 10 above)
Comment (blackyblack) [reply include/veriblock/pop/blockchain/blockchain_util.hpp:131]: Yes it is intentional. We do not want recoverEndorsements to modify dirtyness of the block. So if it was dirty, it will remain dirty. And if it was clean, it will remain clean.

**Review by Warchant — APPROVED** [review]
